File: Task/Discordian-date/Scala/discordian-date.scala
import java.util.{GregorianCalendar, Calendar}

val DISCORDIAN_SEASONS=Array("Chaos", "Discord", "Confusion", "Bureaucracy", "The Aftermath")
def ddate(year:Int, month:Int, day:Int):String={
   val date=new GregorianCalendar(year, month-1, day)
   val dyear=year+1166

   val isLeapYear=date.isLeapYear(year)
   if(isLeapYear && month==2 && day==29)
      return "St. Tib's Day "+dyear+" YOLD"

   var dayOfYear=date.get(Calendar.DAY_OF_YEAR)
   if(isLeapYear && dayOfYear>=60)
      dayOfYear-=1	// compensate for St. Tib's Day

   val dday=dayOfYear%73
   val season=dayOfYear/73
   "%s %d, %d YOLD".format(DISCORDIAN_SEASONS(season), dday, dyear)
}

File: Task/Discordian-date/Perl/discordian-date.pl
use 5.010;
use strict;
use warnings;
use Time::Piece ();

my @seasons = (qw< Chaos Discord Confusion Bureaucracy >, 'The Aftermath');
my @week_days = (qw< Sweetmorn Boomtime Pungenday Prickle-Prickle >, 'Setting Orange');

sub ordinal {
	my ($n) = @_;
	return $n . "th" if int($n/10) == 1;
	return $n . ((qw< th st nd rd th th th th th th>)[$n % 10]);
}

sub ddate {
	my $d = Time::Piece->strptime( $_[0], '%Y-%m-%d' );
	my $yold = 'in the YOLD ' . ($d->year + 1166);

	my $day_of_year0 = $d->day_of_year;

	if( $d->is_leap_year ) {
		return "St. Tib's Day, $yold" if $d->mon == 2 and $d->mday == 29;
		$day_of_year0-- if $day_of_year0 >= 60; # Compensate for St. Tib's Day
	}

	my $weekday = $week_days[ $day_of_year0 % @week_days ];
	my $season = $seasons[ $day_of_year0 / 73 ];
	my $season_day = ordinal( $day_of_year0 % 73 + 1 );

	return "$weekday, the $season_day day of $season $yold";
}

say "$_ is " . ddate($_) for qw< 2010-07-22 2012-02-28 2012-02-29 2012-03-01 >;
